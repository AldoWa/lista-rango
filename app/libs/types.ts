export enum Days {
  DOMINGO = 1,
  SEGUNDA = 2,
  TERCA = 3,
  QUARTA = 4,
  QUINTA = 5,
  SEXTA = 6,
  SABADO = 7
}

interface Hour {
  from: string;
  to: string;
  days: Days[];
}

export interface Restaurant {
  id: number;
  name: string;
  address: string;
  image: string;
  hours: Hour[];
}

export interface RestaurantResponse {
  data: Restaurant[];
}