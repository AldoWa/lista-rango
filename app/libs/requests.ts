import { NextResponse } from "next/server";
import { RestaurantResponse } from "./types";

export const getRestaurants = async (): Promise<RestaurantResponse> => {
  const response = await fetch('http://localhost:3000/api/restaurants') as NextResponse<RestaurantResponse>;
  if (!response.ok) {
    throw new Error('Error in get restaurants')
  }
  return response.json();
}
