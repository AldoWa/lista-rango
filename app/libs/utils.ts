import moment from "moment"
import { Hour } from "./types";

/**
 * 1 - Fazer um map para adicionar se Ã© aberto ou fechado
 * 2 - Comparar o Dia de Hoje com os dias do Array e a Hora
 */

export function isOpened(hours: Hour[]){
  const todayDay = moment().day();
  const verifyTodayIsOpened = hours.some((hour) => {
    return hour.days.includes(todayDay + 1);
  })

  const verifyHourIsOpened = hours.some((hour) => {
    const dateHourFrom = moment(hour.from, "HH:mm")
    const dateHourTo = moment(hour.to, "HH:mm")
    const actualDate = moment()
    if(dateHourTo.isSameOrBefore(dateHourFrom)){
      return actualDate.isSameOrAfter(dateHourFrom) && actualDate.isBefore(dateHourTo.add(1, 'days'))
    }
    return actualDate.isSameOrAfter(dateHourFrom) && actualDate.isBefore(dateHourTo);
  })

  return verifyHourIsOpened && verifyTodayIsOpened;
}