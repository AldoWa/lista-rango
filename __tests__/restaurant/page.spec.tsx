import "@testing-library/jest-dom";
import { screen, render } from "@testing-library/react";
import Page from "@/app/restaurants/[id]/menu/page";
import { Days } from "@/app/libs/types";
jest.mock("@/app/libs/requests", () => {
  return {
    getRestaurant: jest.fn(() =>
      Promise.resolve({
        data: [
          {
            restaurantId: 1,
            name: "Lasanha de frango",
            image:
              "Lasanha image",
            price: 12.0,
            group: "Almoço",
            sales: [
              {
                description: "Promo Almoço",
                price: 6.0,
                hours: [
                  {
                    from: "11:00",
                    to: "12:00",
                    days: [Days.DOMINGO, Days.QUARTA, Days.SABADO],
                  },
                ],
              },
            ],
          },
        ],
      })
    ),
  };
});

describe("Restaurant -> Page", () => {
  it("Should render the page with the restaurant id 1", async () => {
    render(<Page params={{ id: "1" }} />);

    const banner = await screen.findByTestId("banner");
    const search = await screen.findByTestId("search");

    expect(banner).toBeInTheDocument();
    expect(search).toBeInTheDocument();
  });

  it("Should render the page with only one category", async () => {
    render(<Page params={{ id: "1" }} />);
    
    const accordion = await screen.findByTestId("accordion");
    const accordionTitle = await screen.findByTestId("accordion-title");

    expect(accordion).toBeInTheDocument();
    expect(accordionTitle).toBeInTheDocument();
    
    expect(accordionTitle).toHaveTextContent("Almoço");
  })

  it('Should match snapshot', () => {
    const { container } = render(<Page params={{ id: "1" }} />);
    expect(container).toMatchSnapshot()
  })
});
