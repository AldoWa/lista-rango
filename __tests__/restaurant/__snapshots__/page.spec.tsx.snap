// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Restaurant -> Page Should match snapshot 1`] = `
<div>
  <div
    class="inline-flex items-center max-w-3xl mt-6 mb-4"
    data-testid="banner"
  >
    <img
      alt="Picture restaurant"
      data-nimg="1"
      decoding="async"
      height="145"
      loading="lazy"
      src="/_next/image?url=%2Fimg.jpg&w=384&q=100"
      srcset="/_next/image?url=%2Fimg.jpg&w=256&q=100 1x, /_next/image?url=%2Fimg.jpg&w=384&q=100 2x"
      style="color: transparent;"
      width="145"
    />
    <div
      class="flex flex-col gap-y-1"
    >
      <h1
        class="text-2xl font-medium"
      >
        Nome do Restaurante
      </h1>
      <p
        class="font-normal text-base"
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
      </p>
      <ul
        class="flex flex-col gap-y-0.5"
      >
        <li
          class="text-xs"
        >
          <span>
            Segunda à Sexta:
          </span>
           
          <b>
            11:30 às 15:00
          </b>
        </li>
        <li
          class="text-xs"
        >
          <span>
            Sábados:
          </span>
           
          <b>
            11:30 às 22:00
          </b>
        </li>
        <li
          class="text-xs"
        >
          <span>
            Domingos e Feriados:
          </span>
           
          <b>
            11:30 às 15:00
          </b>
        </li>
      </ul>
    </div>
  </div>
  <section
    class="grid grid-cols-[minmax(802px,_1fr)_282px] gap-x-36 max-[1280px]:grid-cols-1"
  >
    <div>
      <div
        class="flex justify-start relative"
        data-testid="search"
      >
        <input
          class=" shadow-md  pr-8 pl-52 w-full  rounded-full  bg-slate-100 h-10  focus:outline-none my-8 max-w-4xl "
          type="text"
        />
        <span
          class="absolute flex items-center rounded-l-full h-10 bottom-8 left-0 bg-white px-7 "
        >
          Buscar no cardápio
        </span>
      </div>
      <p>
        Carregando cardapio...
      </p>
    </div>
    <div
      class="w-64 h-[48rem] bg-gray-600 max-[1280px]:m-6 max-[1280px]:w-96"
    />
  </section>
</div>
`;
